src/frontend/contexts/TrackStoreContext.tsx

```diff
@@ -0,0 +1,186 @@
+import {createContext, useContext} from 'react';
+import {createStore, useStore, type StoreApi} from 'zustand';
+import {
+  createJSONStorage,
+  persist as createPersistedState,
+} from 'zustand/middleware';
+import * as v from 'valibot';
+
+import {MMKVZustandStorage} from '../hooks/persistedState/createPersistedState';
+import {LocationHistoryPoint} from '../sharedTypes/location';
+import {calculateTotalDistance} from '../utils/distance';
+
+const TrackStateSchema = v.intersect([
+  v.object({
+    description: v.string(),
+    distance: v.number(),
+    locationHistory: v.array(
+      // TODO: Create schema for this
+      v.object({
+        latitude: v.number(),
+        longitude: v.number(),
+        timestamp: v.number(),
+      }),
+    ),
+    observationRefs: v.array(
+      // TODO: Create schema for this
+      v.object({
+        docId: v.string(),
+        versionId: v.string(),
+      }),
+    ),
+  }),
+  v.union([
+    v.object({
+      isTracking: v.literal(true),
+      trackingSince: v.date(),
+    }),
+    v.object({
+      isTracking: v.literal(false),
+      trackingSince: v.null(),
+    }),
+  ]),
+]);
+
+export type TrackState = v.InferOutput<typeof TrackStateSchema>;
+
+// NOTE: Do not change!
+const STORAGE_KEY = 'MapeoTrack' as const;
+
+function createInitialState(): TrackState {
+  return {
+    description: '',
+    distance: 0,
+    isTracking: false,
+    locationHistory: [],
+    observationRefs: [],
+    trackingSince: null,
+  };
+}
+
+export function createTrackStore({persist} = {persist: false}) {
+  let store: StoreApi<TrackState>;
+
+  if (persist) {
+    store = createStore(
+      createPersistedState(createInitialState, {
+        name: STORAGE_KEY,
+        storage: createJSONStorage(() => MMKVZustandStorage),
+        version: 1,
+        migrate: (persistedState, version): TrackState => {
+          const newState = createInitialState();
+
+          if (version === 0) {
+            try {
+              // Even if persisted state has fields that are not part of the schema,
+              // Valibot will only extract the relevant fields (assuming all relevant ones pass validation).
+              // https://valibot.dev/guides/objects/
+              return v.parse(TrackStateSchema, persistedState, {
+                abortEarly: true,
+              });
```

Comment:
not in love with having to remember this detail, but it definitely is convenient in this case.